File: dotnet/CardsProject/Domain/Tournaments/Season.cs
using CardsProject.Domain.Players;
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Tournaments;

public enum SeasonFormatType
{
    Standard,
    Extended,
    Legacy,
    Vintage,
    Commander,
    Draft
}

public class Season : IValidatableObject
{
    public int Id { get; set; }

    public string Name { get; set; } = "";
    public DateOnly? StartDate { get; set; } = null;
    public DateOnly? EndDate { get; set; } = null;
    public SeasonFormatType Format { get; set; }
    public bool IsActive { get; set; } = false;
    public string? RewardDescription { get; set; }

    public ICollection<CardsProject.Domain.Players.PlayerSeasonStats> PlayerStats { get; set; } = new List<CardsProject.Domain.Players.PlayerSeasonStats>();
    public ICollection<Tournament> Tournaments { get; set; } = new List<Tournament>();

    // Business operations

    public void Activate()
    {
        // TODO: implement activate
    }

    public void Deactivate()
    {
        // TODO: implement deactivate
    }

    public void FinalizeRewards()
    {
        // TODO: implement finalize_rewards
    }

    public bool IsOngoing()
    {
        // TODO: implement is_ongoing
        return default;
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( (EndDate == null || (StartDate != null && EndDate > StartDate)) ))
            yield return new ValidationResult("Season end date must be after start date", new[] { nameof(Id) });
    }
}

File: rails/app/models/season.rb
class Season < ApplicationRecord
  self.table_name = 'seasons'

  enum :format, { standard: 0, extended: 1, legacy: 2, vintage: 3, commander: 4, draft: 5 }

  has_many :player_stats, class_name: 'PlayerSeasonStats', inverse_of: :season
  has_many :tournaments, class_name: 'Tournament', inverse_of: :season

  validates :name, presence: true, length: { maximum: 200 }

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:end_date_after_start_date, 'Season end date must be after start date') unless ((end_date.nil? || (!start_date.nil? && end_date > start_date)))
  end

  def to_s
    name.to_s
  end

  # Business operations

  def activate
    # TODO: implement activate
  end

  def deactivate
    # TODO: implement deactivate
  end

  def finalize_rewards
    # TODO: implement finalize_rewards
  end

  def is_ongoing
    # TODO: implement is_ongoing
    nil
  end
end
